materialize-mongodb/transactor.go

```diff
@@ -269,8 +269,12 @@ func (t *transactor) storeWorker(
 				return fmt.Errorf("bulk write for collection %s: %w", collection.Name(), err)
 			}
 
-			// Sanity check the result of the bulk write operation.
-			if res.ModifiedCount != expectModified || res.InsertedCount != expectInserted || res.DeletedCount != expectDeleted {
+			// Sanity check the result of the bulk write operation. For updated documents, we check
+			// MatchedCount instead of ModifiedCount since certain Flow reduction strategies can
+			// result in identical documents being stored, and MongoDB does not report an attempted
+			// "replace" with an identical document as an update when it's part of a bulk write
+			// operation.
+			if res.MatchedCount != expectModified || res.InsertedCount != expectInserted || res.DeletedCount != expectDeleted {
```

Comment:
This `MatchedCount` doesn't include matched delete documents? 🤔

Reply:
Apparently not, the description for `MatchedCount` is "The number of documents matched by filters in update and replace operations.". So for our purposes its how many "replace" operations we have.